### PR TITLE
Optionally Reduce LR on plateau

### PR DESCRIPTION
**Pull request by johnlockejrr**
# Learning Rate Warmup and Optimization Implementation

## Overview
Added learning rate warmup functionality to improve training stability, especially when using pretrained weights. The implementation uses TensorFlow's native learning rate scheduling for better performance.

### 1. Configuration Updates (`runs/train_no_patches_448x448.json`)
Added new configuration parameters for warmup:
```json
{
    "warmup_enabled": true,
    "warmup_epochs": 5,
    "warmup_start_lr": 1e-6
}
```

#### A. Optimizer and Learning Rate Schedule
- Replaced fixed learning rate with dynamic scheduling
- Implemented warmup using `tf.keras.optimizers.schedules.PolynomialDecay`
- Maintained compatibility with existing ReduceLROnPlateau and EarlyStopping

```python
if warmup_enabled:
    lr_schedule = tf.keras.optimizers.schedules.PolynomialDecay(
        initial_learning_rate=warmup_start_lr,
        decay_steps=warmup_epochs * steps_per_epoch,
        end_learning_rate=learning_rate,
        power=1.0  # Linear decay
    )
    optimizer = Adam(learning_rate=lr_schedule)
else:
    optimizer = Adam(learning_rate=learning_rate)
```

#### B. Learning Rate Behavior
- Initial learning rate: 1e-6 (configurable via `warmup_start_lr`)
- Target learning rate: 5e-5 (configurable via `learning_rate`)
- Linear increase over 5 epochs (configurable via `warmup_epochs`)
- After warmup, learning rate remains at target value until ReduceLROnPlateau triggers

## Benefits
1. Improved training stability during initial epochs
2. Better handling of pretrained weights
3. Efficient implementation using TensorFlow's native scheduling
4. Configurable through JSON configuration file
5. Maintains compatibility with existing callbacks (ReduceLROnPlateau, EarlyStopping)

## Usage
To enable warmup:
1. Set `warmup_enabled: true` in the configuration file
2. Adjust `warmup_epochs` and `warmup_start_lr` as needed
3. The warmup will automatically integrate with existing learning rate reduction and early stopping

To disable warmup:
- Set `warmup_enabled: false` or remove the warmup parameters from the configuration file

### CODE REVIEW

**Comment by johnlockejrr** [comment]
With `Warmup`, `ReduceLROnPlateau` and `early stopping` the training process looks smoother and much more stable than the default:

```
==================================================================================================
Total params: 38,211,247
Trainable params: 38,154,089
Non-trainable params: 57,158
__________________________________________________________________________________________________
Epoch 1/50
1194/1194 [==============================] - ETA: 0s - loss: 0.7081 - accuracy: 0.7516
Epoch 1: saving model to runs/sam_41_mss_npt_448x448/model_1
INFO - tensorflow - Assets written to: runs/sam_41_mss_npt_448x448/model_1/assets
1194/1194 [==============================] - 396s 302ms/step - loss: 0.7081 - accuracy: 0.7516 - val_loss: 0.5615 - val_accuracy: 0.8679 - lr: 4.7968e-06
Epoch 2/50
1194/1194 [==============================] - ETA: 0s - loss: 0.5898 - accuracy: 0.8990
Epoch 2: saving model to runs/sam_41_mss_npt_448x448/model_2
INFO - tensorflow - Assets written to: runs/sam_41_mss_npt_448x448/model_2/assets
1194/1194 [==============================] - 357s 299ms/step - loss: 0.5898 - accuracy: 0.8990 - val_loss: 0.5170 - val_accuracy: 0.8843 - lr: 8.5968e-06
Epoch 3/50
1194/1194 [==============================] - ETA: 0s - loss: 0.5065 - accuracy: 0.9154
Epoch 3: saving model to runs/sam_41_mss_npt_448x448/model_3
INFO - tensorflow - Assets written to: runs/sam_41_mss_npt_448x448/model_3/assets
1194/1194 [==============================] - 356s 298ms/step - loss: 0.5065 - accuracy: 0.9154 - val_loss: 0.4268 - val_accuracy: 0.8995 - lr: 1.2397e-05
Epoch 4/50
1194/1194 [==============================] - ETA: 0s - loss: 0.4244 - accuracy: 0.9236
Epoch 4: saving model to runs/sam_41_mss_npt_448x448/model_4
INFO - tensorflow - Assets written to: runs/sam_41_mss_npt_448x448/model_4/assets
1194/1194 [==============================] - 357s 299ms/step - loss: 0.4244 - accuracy: 0.9236 - val_loss: 0.3657 - val_accuracy: 0.9017 - lr: 1.6197e-05
Epoch 5/50
1194/1194 [==============================] - ETA: 0s - loss: 0.3661 - accuracy: 0.9287
Epoch 5: saving model to runs/sam_41_mss_npt_448x448/model_5
INFO - tensorflow - Assets written to: runs/sam_41_mss_npt_448x448/model_5/assets
1194/1194 [==============================] - 358s 299ms/step - loss: 0.3661 - accuracy: 0.9287 - val_loss: 0.3451 - val_accuracy: 0.9161 - lr: 1.9997e-05
Epoch 6/50
1194/1194 [==============================] - ETA: 0s - loss: 0.3299 - accuracy: 0.9324
Epoch 6: saving model to runs/sam_41_mss_npt_448x448/model_6
INFO - tensorflow - Assets written to: runs/sam_41_mss_npt_448x448/model_6/assets
1194/1194 [==============================] - 356s 298ms/step - loss: 0.3299 - accuracy: 0.9324 - val_loss: 0.3256 - val_accuracy: 0.9052 - lr: 2.0000e-05
Epoch 7/50
1194/1194 [==============================] - ETA: 0s - loss: 0.3069 - accuracy: 0.9355
Epoch 7: saving model to runs/sam_41_mss_npt_448x448/model_7
INFO - tensorflow - Assets written to: runs/sam_41_mss_npt_448x448/model_7/assets
1194/1194 [==============================] - 357s 299ms/step - loss: 0.3069 - accuracy: 0.9355 - val_loss: 0.2609 - val_accuracy: 0.9106 - lr: 2.0000e-05
Epoch 8/50
1194/1194 [==============================] - ETA: 0s - loss: 0.2906 - accuracy: 0.9380
Epoch 8: saving model to runs/sam_41_mss_npt_448x448/model_8
INFO - tensorflow - Assets written to: runs/sam_41_mss_npt_448x448/model_8/assets
1194/1194 [==============================] - 356s 298ms/step - loss: 0.2906 - accuracy: 0.9380 - val_loss: 0.2852 - val_accuracy: 0.9165 - lr: 2.0000e-05
Epoch 9/50
1194/1194 [==============================] - ETA: 0s - loss: 0.2774 - accuracy: 0.9403
Epoch 9: saving model to runs/sam_41_mss_npt_448x448/model_9
INFO - tensorflow - Assets written to: runs/sam_41_mss_npt_448x448/model_9/assets
1194/1194 [==============================] - 358s 300ms/step - loss: 0.2774 - accuracy: 0.9403 - val_loss: 0.2608 - val_accuracy: 0.9229 - lr: 2.0000e-05
Epoch 10/50
1194/1194 [==============================] - ETA: 0s - loss: 0.2656 - accuracy: 0.9422
Epoch 10: saving model to runs/sam_41_mss_npt_448x448/model_10
INFO - tensorflow - Assets written to: runs/sam_41_mss_npt_448x448/model_10/assets
1194/1194 [==============================] - 357s 299ms/step - loss: 0.2656 - accuracy: 0.9422 - val_loss: 0.2218 - val_accuracy: 0.9341 - lr: 2.0000e-05
Epoch 11/50
 254/1194 [=====>........................] - ETA: 4:26 - loss: 0.2586 - accuracy: 0.9439
```

We'll see at training's end.
